[Hot Corners] Added hover delay functionality.

Odyseus opened the pull request:
This pull request adds the ability to set a hover delay to activate hot corners. Its main purpose is to 
avoid accidental hot corners activation. I also added tooltips to the hot corners icons that reflect their current action, as suggested by @anandrkris.

The new Hot Corners settings window would look like this:

![hotcornershoverdelayimplemented2](https://cloud.githubusercontent.com/assets/3822556/19623063/73cd232a-9891-11e6-8504-51f4bec855cb.png)

### Some technicalities
- File js/ui/hotCorner.js 
    - This PR also fixes some minor white-space *inconsistencies* on this file (some missing spaces and some extra tabs). But, as I use a code formatter plugin, it also expanded some objects. Let me know if I should revert back those objects expansions.
- File files/usr/share/cinnamon/cinnamon-settings/modules/cs_hotcorner.py
    - Here I had to add a timer to the **on_widget_changed** function to avoid repeated triggering from the new **SpinButton** elements. I tried to add that timer to a new callback for the **SpinButton** elements, but I wasn't able to pass the correct arguments to the **on_widget_changed** function. Any advice will be welcome.
    - <del>I used a **Gtk.Grid** instead of a **Gtk.HBox** because of its deprecation.</del> Switched to a **Gtk.Box** as suggested by @collinss.
- <del>This PR also adds a new string that needs to be translated. I will need advice on how to proceed with this as well.</del>

anandrkris commented:
Hope this gets merged - lot of votes for this :)
https://github.com/linuxmint/Cinnamon/issues/1050

anandrkris commented:
Also, a suggestion - when icon is visible in the corner, can the action be displayed as a tooltip on hover over icon?
Show All workspaces, windows, custom action, etc

Odyseus commented:
Hello, @anandrkris.

I like your suggestion about adding a tooltip to the hot corners icons. I have been doing some tests but, sadly, without success. No matter how or at which point I add the tooltip, it never shows on icon hover.

Here is a shortened version of what I did in case someone has any suggestion:

After adding the import for the tooltip module in the **js/ui/hotCorner.js** file...

```javascript
const Tooltips = imports.ui.tooltips;
```

...I simply added the tooltip to the icon inside the **setProperties** method like this...

```javascript
if (this.icon) {
    new Tooltips.Tooltip(this.iconActor, "Tooltip text.");
    this.iconActor.show();
} else
    this.iconActor.hide();
```

Is it even possible for the hot corners icons to have a tooltip?

NikoKrause commented:
@Odyseus 
Try this.
```js 
this.tooltip = new Tooltips.Tooltip(this.iconActor, "Tooltip text.");
```

Odyseus commented:
Hello, @NikoKrause.

I already tried that. The very first thing that I tried was to create the tooltip inside **HotCorner._init** and then use the **set_text()** method inside the **setProperties** method to apply the correct tooltip text. It's pretty straight forward and it works in every applet code that I ever used, except here with the hot corners.

NikoKrause commented:
@Odyseus
Does my suggestion doesn't work?
I tried your code and added `this.tooltip =`, which you forgot before `new`
And it showed the Tooltip on icon hover like expected.

Odyseus commented:
@NikoKrause.

> Does my suggestion doesn't work?

Like I said, I already tried that. I used the exact code that you posted and it didn't work.

> I tried your code and added this.tooltip =, which you forgot before new

I didn't really forget. It isn't needed to store the tooltip in a new variable because the tooltip is already *added* to the element passed by the first argument of the constructor ([here](https://github.com/Odyseus/CinnamonTools/blob/master/Applets/0dyseus%40QuickMenu/applet.js#L375) is a functional example that I use in one of my applets). Eighter way, like I stated before, I already added `this.tooltip = ` in front of that line and it still doesn't work for me.

I did my tests in all of my 4 Cinnamon systems with the same results. The tooltip isn't shown when hovering over the hot corner icon.

I will give this rest for now. Maybe someone else can create a different pull request that adds the tooltips to the hot corners icons.

NikoKrause commented:
@Odyseus 
Could you check if my branch works for you https://github.com/NikoKrause/Cinnamon/tree/hot-corner-tooltip? I created a commit with your code in there https://github.com/NikoKrause/Cinnamon/commit/60993434fb7410879a943358c02ce0ff334cb2a7.

Just replace the hotCorner.js file in `/usr/share/cinnamon/js/ui` with the file from my branch.

It looks like this on my machine:
![bildschirmfoto vom 2016-10-25 20-25-10](https://cloud.githubusercontent.com/assets/8415124/19698860/3753f7b8-9af1-11e6-99d5-cecf422537f5.png)

Odyseus commented:
@NikoKrause

No, it still isn't working on any of the 4 Cinnamon systems I tried before. **But**, I tried it on a fith system (an Archlinux system with Cinnamon) and all tooltips work like a charm here! This is very strange!

I hesitate to add to this pull request something that I can't make work reliable on all my systems. If you want to add this feature yourself and send the pull request, [here is a Gist](https://gist.github.com/Odyseus/084193105d6bd8fc9cc06da4e4f4a237) with the **hotCorner.js** file with icons tooltips fully implemented.

Odyseus commented:
@NikoKrause and other code reviewers.

Ignore the previous posts discussing problems while setting the tooltips for the hot corners icons. I had an extension enabled that was interfering with the hot corners code. I disabled it and everything started working as it should.

@anandrkris: If this PR *doesn't make it*, NikoKrause just send a [pull request](https://github.com/linuxmint/Cinnamon/pull/5852) adding the tooltips to the hot corner icons.

Apologies for the *noise*.

collinss commented:
@Odyseus A couple notes/suggestions.

- You don't need to worry about the translations. Putting ```_()``` around the string is enough to insure that it will get translated for the next release.
- ```Gtk.HBox``` and ```Gtk.VBox``` were deprecated in favor of ```Gtk.Box```, which adds an additional orientation property to make it vertical. Unless you need the extra layout features of ```Gtk.Grid``` (which you do not here), I would recommend using ```Gtk.Box```.
- I would recommend adding the tooltip to the actor as soon as you create it. If you need to make sure the tooltip is not shown under certain circumstances, you can set the text to an empty string later (you don't need to do this if your actor is hidden as it wont show anyway). The way you did it above could lead to multiple tooltips on the same actor, all competing with each other. This might also solve your problem with it not showing, but I can't guarantee anything there.

Odyseus commented:
Hello, @collinss.

- **About translations:** Got it.
- **About `Gtk.Box`s:** I have chosen a `Gtk.Grid` because I couldn't make the `Gtk.Box` *"look good"* (basically, I couldn't understand the arguments passed to **pack_start**). But since you recommended me to use a `Gtk.Box`, I gave it another try and I successfully made it work.
- **About the tooltips:** The code that I posted was just an example. So that someone willing to help me out finding the root of my problem could quickly set up a tooltip with just two lines of code. Thanks to NikoKrause's assistance, I figured that the *problem* was on several of my systems.

Thanks for your notes/suggestions, collinss.

Pull request updated.

anandrkris commented:
Thanks @Odyseus for listening :smile: and again for your efforts.  :clap:
